# noom 데이터 분석 수행

In [1]:
import pandas as pd
import numpy as np
import os

from IPython.core.display import display, HTML
display(HTML("<style>.container { width:100% !important; }</style>"))

#  0. 데이터 로딩하기

In [2]:
raw_data = pd.read_csv(os.getcwd() + '/noom/noom_user.csv', parse_dates=["Purchased At"]) # 사용자의 프로필(성별, 나이 등)과 구매 정보 등
print(raw_data.shape)
raw_data.head()

(10000, 14)


Access Code Name  Gender   Age  Height  Initial Weight  Lowest Weight  \
0    Y9RY2VSI  김승혜  FEMALE  25.0   172.0            66.9           65.8   
1    3GTN3S3B  허승준    MALE  26.0   176.0            70.0            NaN   
2    6B0IG276  이지민  FEMALE  23.0   171.0            98.0            NaN   
3    EMGRU2MO  장설윤  FEMALE  20.0   160.0            70.7            NaN   
4    1ELG96TX  서성빈  FEMALE  28.0   165.0            55.5            NaN   

   Target Weight         Product Name     Status   Price  \
0      55.000000          눔 체중감량 프로그램  completed  112500   
1      65.000000          눔 체중감량 프로그램  completed   44780   
2      91.140000  눔 체중감량 프로그램 (천원 체험)  completed  132000   
3      53.000000  눔 체중감량 프로그램 (천원 체험)  completed  112500   
4      51.615002          눔 체중감량 프로그램  completed   44780   

             Purchased At Payment Type   Channel  
0 2017-04-14 19:03:29.976    Recurring    others  
1 2017-05-23 20:53:54.368    Recurring    others  
2 2017-08-23 23:39:21.840    Recurring  facebook  
3 2017-08-28 20:18:22.824    Recurring     naver  
4 2017-05-07 17:50:30.944    Recurring  facebook

In [3]:
coach = pd.read_csv(os.getcwd() + '/noom/noom_coach.csv') # 값은 코치가 사용자에게 코칭을 한 횟수
print(coach.shape)
coach.head()

(10000, 101)


Access Code  정은오 코치(VEV4PGJB)  오승혁 코치(VENPKBP9)  조소은 코치(D0WASBXN)  \
0    Y9RY2VSI                 0                 0                 0   
1    3GTN3S3B                 0                 0                 0   
2    6B0IG276                 0                 0                 0   
3    EMGRU2MO                 0                 0                 0   
4    1ELG96TX                 0                 0                 0   

   고영재 코치(C91AVNGB)  조수민 코치(OBCAO3W0)  강채아 코치(WH2NIKCO)  황다훈 코치(1I6IWURH)  \
0                 0                 0                 0                 0   
1                 0                 0                 0                 0   
2                 0                 0                 0                 0   
3                 0                 0                 0                 1   
4                 0                 0                 0                 0   

   백슬은 코치(228BFB50)  유채우 코치(IW53Y9AW)  ...  오초빈 코치(A3WOLAQM)  \
0                 0                 0  ...                 0   
1                 0                 0  ...                 0   
2                 0                 0  ...                 0   
3                 0                 0  ...                 0   
4                 0                 0  ...                 0   

   서수정 코치(F36LORFC)  정서율 코치(LX1G7EMD)  고우재 코치(SKNL9Z4P)  문한규 코치(OU1WVDGA)  \
0                 0                 0                 1                 0   
1                 0                 0                 0                 1   
2                 0                 0                 0                 0   
3                 0                 0                 0                 0   
4                 0                 0                 0                 0   

   황세안 코치(3QUBQAVE)  홍성은 코치(2I3QJQ5O)  고성은 코치(34T7XPYR)  백한율 코치(HPWAN8R0)  \
0                 0                 0                 0                 0   
1                 0                 0                 0                 0   
2                 0                 0                 0                 0   
3                 0                 0                 0                 0   
4                 0                 0                 0                 0   

   안슬은 코치(QAVWJSZ1)  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0  

[5 rows x 101 columns]

----
# (1) 데이터 정리

## 1. 전체 컬럼에서 필요한 컬럼만 가져오세요.
>- Access Code(index지정), Name, Gender, Age, Height, Initial Weight, Lowest Weight, Target Weight
>  , Status, Price, Purchased At, Channel

In [4]:
data = raw_data[['Access Code', 'Name', 'Gender', 'Age', 'Height', 'Initial Weight', 
             'Lowest Weight', 'Target Weight', 'Status', 'Price', 'Purchased At', 'Channel']].set_index('Access Code')
data.head()

Name  Gender   Age  Height  Initial Weight  Lowest Weight  \
Access Code                                                             
Y9RY2VSI     김승혜  FEMALE  25.0   172.0            66.9           65.8   
3GTN3S3B     허승준    MALE  26.0   176.0            70.0            NaN   
6B0IG276     이지민  FEMALE  23.0   171.0            98.0            NaN   
EMGRU2MO     장설윤  FEMALE  20.0   160.0            70.7            NaN   
1ELG96TX     서성빈  FEMALE  28.0   165.0            55.5            NaN   

             Target Weight     Status   Price            Purchased At  \
Access Code                                                             
Y9RY2VSI         55.000000  completed  112500 2017-04-14 19:03:29.976   
3GTN3S3B         65.000000  completed   44780 2017-05-23 20:53:54.368   
6B0IG276         91.140000  completed  132000 2017-08-23 23:39:21.840   
EMGRU2MO         53.000000  completed  112500 2017-08-28 20:18:22.824   
1ELG96TX         51.615002  completed   44780 2017-05-07 17:50:30.944   

              Channel  
Access Code            
Y9RY2VSI       others  
3GTN3S3B       others  
6B0IG276     facebook  
EMGRU2MO        naver  
1ELG96TX     facebook


## 2. 성별 컬럼을 정리해주세요.
>- 남성과 여성을 굳이 대문자로 사용할 필요는 없을 것으로 보입니다. 가독성을 높이기 위해, 대문자로 되어있는 텍스트를 소문자로 변경해주세요.
>- 또한 Gender 컬럼을 정리한 뒤, 전체 가입자중 남성의 비율과 여성의 비율을 계산해주세요.

In [5]:
data['Gender'].value_counts(dropna=False)

FEMALE    8846
MALE      1023
NaN        131
Name: Gender, dtype: int64

In [6]:
data.loc[data['Gender'] == 'FEMALE', 'Gender(clean)'] = 'female'
data.loc[data['Gender'] == 'MALE', 'Gender(clean)'] = 'male'
data.head()

Name  Gender   Age  Height  Initial Weight  Lowest Weight  \
Access Code                                                             
Y9RY2VSI     김승혜  FEMALE  25.0   172.0            66.9           65.8   
3GTN3S3B     허승준    MALE  26.0   176.0            70.0            NaN   
6B0IG276     이지민  FEMALE  23.0   171.0            98.0            NaN   
EMGRU2MO     장설윤  FEMALE  20.0   160.0            70.7            NaN   
1ELG96TX     서성빈  FEMALE  28.0   165.0            55.5            NaN   

             Target Weight     Status   Price            Purchased At  \
Access Code                                                             
Y9RY2VSI         55.000000  completed  112500 2017-04-14 19:03:29.976   
3GTN3S3B         65.000000  completed   44780 2017-05-23 20:53:54.368   
6B0IG276         91.140000  completed  132000 2017-08-23 23:39:21.840   
EMGRU2MO         53.000000  completed  112500 2017-08-28 20:18:22.824   
1ELG96TX         51.615002  completed   44780 2017-05-07 17:50:30.944   

              Channel Gender(clean)  
Access Code                          
Y9RY2VSI       others        female  
3GTN3S3B       others          male  
6B0IG276     facebook        female  
EMGRU2MO        naver        female  
1ELG96TX     facebook        female

In [7]:
# 남여 비율
print(data['Gender(clean)'].value_counts())
print(data['Gender(clean)'].value_counts(normalize=True))

female    8846
male      1023
Name: Gender(clean), dtype: int64
female    0.896342
male      0.103658
Name: Gender(clean), dtype: float64



## 3. 키(Height) 컬럼을 정리해주세요.
>- 키가 -1 cm인 사람은 NaN으로 데이터를 넣어주세요.
>- 1) 최소/최대/평균 키를 구하고, 2) 남성/여성별 평균 키를 구해주세요.

In [8]:
data['Height'].min()

-1.0

In [9]:
data[data['Height'] < 100]['Height'].value_counts()

-1.0    21
Name: Height, dtype: int64

In [10]:
data['Height(clean)'] = [np.nan if x == -1 else x for x in data['Height']]
data.head()

Name  Gender   Age  Height  Initial Weight  Lowest Weight  \
Access Code                                                             
Y9RY2VSI     김승혜  FEMALE  25.0   172.0            66.9           65.8   
3GTN3S3B     허승준    MALE  26.0   176.0            70.0            NaN   
6B0IG276     이지민  FEMALE  23.0   171.0            98.0            NaN   
EMGRU2MO     장설윤  FEMALE  20.0   160.0            70.7            NaN   
1ELG96TX     서성빈  FEMALE  28.0   165.0            55.5            NaN   

             Target Weight     Status   Price            Purchased At  \
Access Code                                                             
Y9RY2VSI         55.000000  completed  112500 2017-04-14 19:03:29.976   
3GTN3S3B         65.000000  completed   44780 2017-05-23 20:53:54.368   
6B0IG276         91.140000  completed  132000 2017-08-23 23:39:21.840   
EMGRU2MO         53.000000  completed  112500 2017-08-28 20:18:22.824   
1ELG96TX         51.615002  completed   44780 2017-05-07 17:50:30.944   

              Channel Gender(clean)  Height(clean)  
Access Code                                         
Y9RY2VSI       others        female          172.0  
3GTN3S3B       others          male          176.0  
6B0IG276     facebook        female          171.0  
EMGRU2MO        naver        female          160.0  
1ELG96TX     facebook        female          165.0

In [11]:
data[data['Height'] < 100].head()

Name  Gender   Age  Height  Initial Weight  Lowest Weight  \
Access Code                                                             
O4OWMJG7     오세윤  FEMALE  23.0    -1.0            58.0           57.0   
H6EV5AXL     박슬지  FEMALE  22.0    -1.0            62.9           49.0   
O1IAZS7A     고솔윤  FEMALE  22.0    -1.0            53.0           52.4   
5NEQOWHW     손초영  FEMALE  30.0    -1.0            58.2           58.2   
OFAXUNXD     백채우  FEMALE  22.0    -1.0            71.0           71.0   

             Target Weight     Status   Price            Purchased At  \
Access Code                                                             
O4OWMJG7              47.0  cancelled  112500 2017-07-20 20:20:03.880   
H6EV5AXL              53.0  completed  112500 2017-08-26 15:58:12.208   
O1IAZS7A              46.0  completed  112500 2017-07-09 17:18:54.632   
5NEQOWHW              53.0  cancelled  112500 2017-08-11 16:02:53.216   
OFAXUNXD              57.0  cancelled   44780 2017-05-15 00:52:33.920   

              Channel Gender(clean)  Height(clean)  
Access Code                                         
O4OWMJG7     facebook        female            NaN  
H6EV5AXL     facebook        female            NaN  
O1IAZS7A     facebook        female            NaN  
5NEQOWHW     facebook        female            NaN  
OFAXUNXD     facebook        female            NaN

In [12]:
# 1) 최소/최대/평균 키
data['Height(clean)'].min(), data['Height(clean)'].max(), data['Height(clean)'].mean()

(106.0, 203.2, 163.54161860276196)

In [13]:
# 2) 남성/여성별 평균 키
pd.pivot_table(data, index='Gender(clean)', values='Height(clean)', aggfunc='mean')

Height(clean)
Gender(clean)               
female            162.116913
male              175.831965


## 4. 나이 컬럼을 정리해주세요.
>- 나이가 0인 데이터는 NaN으로 변경
>- 나이가 60세 이상인 데이터는 NaN으로 변경
>- 1) 최소/최대/평균 나이를 구하고, 2) 남성/여성별 평균 나이를 구해주세요.

In [14]:
data['Age'].min(), data['Age'].max()

(0.0, 173.0)

In [15]:
data['Age(clean)'] = [np.nan if (x == 0) | (x >= 60) else x for x in data['Age']]

In [16]:
# 1) 최소/최대/평균 나이
data['Age(clean)'].min(), data['Age(clean)'].max(), data['Age(clean)'].mean()

(13.0, 59.0, 27.39381024860477)

In [17]:
# 2) 남성/여성별 평균 나이
pd.pivot_table(data, index='Gender(clean)', values='Age(clean)', aggfunc='mean')

Age(clean)
Gender(clean)            
female          27.172929
male            29.309127

----
# (2) VIP 구하기 - 운영팀 요청

>- 1) 유료 사용자 중, 잘못된 정보를 기입한 사용자(invalid user)와, 2) 유료 사용자 중, 눔 코치를 사용하여 큰 성과(ex: 몸무게 감량)를 본 사용자(VIP user)를 찾아보도록 하겠습니다.

>- 1번 사용자의 경우, 유료 결제를 했으나 사용자 정보(나이, 키, 몸무게 등)가 잘못 기입되어있다면 담당 코치가 정확한 코칭을 제공해 줄 수 없는 문제가 있습니다. 그러므로 운영팀은 가능한 빠르게 정보를 잘못 기입한 고객을 찾아서 다시 기입해달라고 요청할 필요가 있습니다.
>- 2번 사용자의 경우, 눔의 VIP 사용자로서 추가 혜택(ex: 서비스 무료 이용)을 제공해주는 것을 조건으로, 눔 코치를 대표하는 홍보 모델로서 활동해줄 것을 요청할 수 있습니다.

>- 특히나 다이어트 관련 서비스는 VIP 사용자의 Before / After를 보여주는 것 만큼 좋은 홍보 수단은 없습니다. 그러므로 데이터 분석 팀에서 특정 조건(ex: 10kg 이상 감량 성공)에 만족하는 코어 사용자를 찾아내는 것을 중요합니다.


## 5. 전체 컬럼에서 필요한 컬럼만 가져오세요.
>- Name
,Age(clean)
,Height(clean)
,Initial Weight
,Lowest Weight
,Target Weight
,Status

In [18]:
data2 = data[['Name' ,'Age(clean)' ,'Height(clean)' ,'Initial Weight' ,'Lowest Weight' ,'Target Weight' ,'Status']]
data2.head()

Name  Age(clean)  Height(clean)  Initial Weight  Lowest Weight  \
Access Code                                                                  
Y9RY2VSI     김승혜        25.0          172.0            66.9           65.8   
3GTN3S3B     허승준        26.0          176.0            70.0            NaN   
6B0IG276     이지민        23.0          171.0            98.0            NaN   
EMGRU2MO     장설윤        20.0          160.0            70.7            NaN   
1ELG96TX     서성빈        28.0          165.0            55.5            NaN   

             Target Weight     Status  
Access Code                            
Y9RY2VSI         55.000000  completed  
3GTN3S3B         65.000000  completed  
6B0IG276         91.140000  completed  
EMGRU2MO         53.000000  completed  
1ELG96TX         51.615002  completed


## 6. 주어진 컬럼으로 다음의 추가 정보를 계산해주세요.
>- Weight Loss(goal) - 목표 감량치. Initial Weight 컬럼과 Target Weight의 차이를 나타냅니다. (마이너스가 나올 수 있습니다)
>- Weight Loss(current) - 최대 감량치. Initial Weight 컬럼과 Lowest Weight의 차이를 나타냅니다. (마이너스가 나올 수 있습니다)
>- 체질량지수(BMI) - 키(Height(clean))와 체중(Initial Weight)으로 체지방의 양을 추정하는 공식입니다. (BMI=체중(kg) / (키(m)×키(m)) )
>- PS) 주의: BMI공식에서 사용하는 키는 센치미터(cm)가 아닌 미터(m)라는것에 주의해주세요.

In [19]:
data2['Weight Loss(goal)'] = data2['Initial Weight'] - data2['Target Weight']
data2['Weight Loss(current)'] = data2['Initial Weight'] - data2['Lowest Weight']
data2['BMI'] = data2['Initial Weight'] / ( (data2['Height(clean)']*0.01)*(data2['Height(clean)']*0.01) )
data2.head()

/home/hyunji/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/home/hyunji/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/home/hyunji/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value ins

Name  Age(clean)  Height(clean)  Initial Weight  Lowest Weight  \
Access Code                                                                  
Y9RY2VSI     김승혜        25.0          172.0            66.9           65.8   
3GTN3S3B     허승준        26.0          176.0            70.0            NaN   
6B0IG276     이지민        23.0          171.0            98.0            NaN   
EMGRU2MO     장설윤        20.0          160.0            70.7            NaN   
1ELG96TX     서성빈        28.0          165.0            55.5            NaN   

             Target Weight     Status  Weight Loss(goal)  \
Access Code                                                
Y9RY2VSI         55.000000  completed          11.900000   
3GTN3S3B         65.000000  completed           5.000000   
6B0IG276         91.140000  completed           6.860000   
EMGRU2MO         53.000000  completed          17.700000   
1ELG96TX         51.615002  completed           3.884998   

             Weight Loss(current)        BMI  
Access Code                                   
Y9RY2VSI                      1.1  22.613575  
3GTN3S3B                      NaN  22.598140  
6B0IG276                      NaN  33.514586  
EMGRU2MO                      NaN  27.617187  
1ELG96TX                      NaN  20.385675


## 7. 잘못된 정보를 기입한 사용자(invalid user)를 찾기
>- 필수 (다음의 조건을 만족하지 않으면 Invalid값에는 언제나 False가 들어갑니다)
>    * 눔의 프로그램을 결제한 구매자. (Status == "completed")
>- 옵션 (다음의 조건 중 하나라도 만족할 경우 Invalid값에 True가 들어가야 합니다)
>    * 나이(Age(clean)), 키(Height(clean))와 몸무게(Initial Weight, Lowest Weight, Target Weight) 중 어느 하나라도 NaN이 들어가 있는 경우.
>    * 키를 너무 작게 기입했거나(140cm 미만)나, 정 반대로 너무 크게(200cm 초과) 기입한 사용자.
>    * BMI수치가 너무 낮거나(18.5 미만) 너무 높은 사용자. (30.0 초과)
>    * 목표 감량치(Weight Loss(goal))가 마이너스인 경우. (보통 현재 체중보다 목표 체중을 낮게 설정합니다)

In [20]:
def invalid_user(data):
    
    if (data['Status'] == 'completed') & (
        pd.isnull(data['Age(clean)']) | 
        pd.isnull(data['Height(clean)']) | 
        pd.isnull(data['Weight Loss(goal)']) | 
        pd.isnull(data['Weight Loss(current)']) |
        ( (data['Height(clean)'] < 140) | (data['Height(clean)'] > 200) ) |  # 키
        ( (data['BMI'] < 18.5) | (data['BMI'] > 30) ) |                      # BMI
        ( data['Weight Loss(goal)'] < 0 )):                                  # 목표감량
        
        return True
    
    else:
        return False

In [21]:
data2['invalid'] = data2.apply(lambda x: invalid_user(x), axis=1)
data2.head()

/home/hyunji/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


Name  Age(clean)  Height(clean)  Initial Weight  Lowest Weight  \
Access Code                                                                  
Y9RY2VSI     김승혜        25.0          172.0            66.9           65.8   
3GTN3S3B     허승준        26.0          176.0            70.0            NaN   
6B0IG276     이지민        23.0          171.0            98.0            NaN   
EMGRU2MO     장설윤        20.0          160.0            70.7            NaN   
1ELG96TX     서성빈        28.0          165.0            55.5            NaN   

             Target Weight     Status  Weight Loss(goal)  \
Access Code                                                
Y9RY2VSI         55.000000  completed          11.900000   
3GTN3S3B         65.000000  completed           5.000000   
6B0IG276         91.140000  completed           6.860000   
EMGRU2MO         53.000000  completed          17.700000   
1ELG96TX         51.615002  completed           3.884998   

             Weight Loss(current)        BMI  invalid  
Access Code                                            
Y9RY2VSI                      1.1  22.613575    False  
3GTN3S3B                      NaN  22.598140     True  
6B0IG276                      NaN  33.514586     True  
EMGRU2MO                      NaN  27.617187     True  
1ELG96TX                      NaN  20.385675     True


## 8. VIP 사용자 체크하기
>- 눔의 프로그램을 결제한 구매자. (Status == "completed")
>- 목표 감량치(Weight Loss(goal)), 최종 감량치(Weight Loss(current)), BMI 수치 모두 NaN이 아닌 값이 들어가 있는 사용자.
>- 최종 감량치(Weight Loss(current))가 10kg 이상.
>- BMI 수치가 높은 사용자. (30.0 이상)
>- 최종 감량치(Weight Loss(current))가 목표 감량치(Weight Loss(goal))보다 큰 경우. (다이어트에 성공한 사람)

In [22]:
def vip_user(data):
    
    if (data['Status'] == 'completed') & (
        ( pd.notnull(data['Weight Loss(goal)']) & pd.notnull(data['Weight Loss(current)']) & pd.notnull(data['BMI']) ) &
        ( data['Weight Loss(current)'] >= 10 ) &
        ( data['BMI'] >= 30 ) &
        ( data['Weight Loss(current)'] > data['Weight Loss(goal)'] )):
        
        return True
        
    else: 
        return False

In [23]:
data2['vip'] = data2.apply(lambda x: vip_user(x), axis=1)
data2[data2['vip'] == True]

/home/hyunji/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


Name  Age(clean)  Height(clean)  Initial Weight  Lowest Weight  \
Access Code                                                                  
3T1I8I8E     임솔지        23.0          158.0         80.0137           54.5   
PJYKU9OW     홍윤오        31.0          174.0         99.9000           76.6   
0EMTSGLJ     류선정        34.0          167.0         86.0000           73.2   
FBEAIFW0     서서원        23.0          170.0         95.0000           75.7   
8QQV2YDW     홍서율        23.0          170.0         95.0000           75.7   
99KOLRU8     고서연        26.0          162.0        106.0000           95.2   
IBOWZ9WZ     손서애        22.0          166.0         83.0000           66.1   
6EH2LGR5     문세영        34.0          164.0         88.8000           56.3   
QQLYGTWD     황수윤        26.0          165.0        105.0000           87.8   
4Z1WB3UZ     허지예        26.0          162.0        106.0000           95.2   
2YAKET8R     서승희        30.0          164.0         99.8000           86.4   
LDPPDM0M     윤지안        19.0          168.0         93.1000           65.5   
PKHJWII8     정선영        23.0          170.0         95.0000           75.7   
3B3WQA4A     홍슬비        28.0          158.0         84.0000           69.3   
SDY4VS0P     오채현         NaN          170.0        103.1000           57.1   

             Target Weight     Status  Weight Loss(goal)  \
Access Code                                                
3T1I8I8E         77.745742  completed           2.267958   
PJYKU9OW         84.000000  completed          15.900000   
0EMTSGLJ         80.050003  completed           5.949997   
FBEAIFW0         85.000000  completed          10.000000   
8QQV2YDW         85.000000  completed          10.000000   
99KOLRU8         98.580000  completed           7.420000   
IBOWZ9WZ         73.000000  completed          10.000000   
6EH2LGR5         82.584003  completed           6.215997   
QQLYGTWD         97.650000  completed           7.350000   
4Z1WB3UZ         98.580000  completed           7.420000   
2YAKET8R         92.814000  completed           6.986000   
LDPPDM0M         69.000000  completed          24.100000   
PKHJWII8         85.000000  completed          10.000000   
3B3WQA4A         75.000000  completed           9.000000   
SDY4VS0P        100.099998  completed           3.000001   

             Weight Loss(current)        BMI  invalid   vip  
Access Code                                                  
3T1I8I8E                  25.5137  32.051634     True  True  
PJYKU9OW                  23.3000  32.996433     True  True  
0EMTSGLJ                  12.8000  30.836531     True  True  
FBEAIFW0                  19.3000  32.871972     True  True  
8QQV2YDW                  19.3000  32.871972     True  True  
99KOLRU8                  10.8000  40.390184     True  True  
IBOWZ9WZ                  16.9000  30.120482     True  True  
6EH2LGR5                  32.5000  33.016062     True  True  
QQLYGTWD                  17.2000  38.567493     True  True  
4Z1WB3UZ                  10.8000  40.390184     True  True  
2YAKET8R                  13.4000  37.105889     True  True  
LDPPDM0M                  27.6000  32.986111     True  True  
PKHJWII8                  19.3000  32.871972     True  True  
3B3WQA4A                  14.7000  33.648454     True  True  
SDY4VS0P                  46.0000  35.674740     True  True

----
# (2) 결제 체크 - 마케팅 요청
>- 가장 중요시 여기는 지표
>   * 한 명의 고객을 데려오는데 필요한 비용, 줄여서 고객 획득 비용(Customer Acquision Cost, 이하 CAC)
>   * 한 명의 고객을 데려왔을 때, 고객이 회사에게 제공해주는 수익(Customer Lifetime Value, 이하 LTV)
>   * 눔 코치에 헌신하는 모든 팀은 LTV를 최대한 높이고, 동시에 CAC를 최대한 낮추는 쪽으로 서비스를 개선합니다.

>- 마케팅팀이 데이터분석가에게 요청하는 내용은 다음과 같습니다.
>   * LTV가 높은 고객군의 인구통계학적 정보. 가령 눔 코치와 같은 다이어트 서비스에서는 남성보다 여성이 서비스의 만족도가 높고 많은 지출을 할 가능성이 있습니다. 이 경우, 페이스북 마케팅을 할 때 여성 고객들에게 집중적으로 광고를 보여주도록 타게팅 할 수 있습니다.
>   * 요일/시간별 결제 비율. 가령 주중보다 주말에 결제할 확률이 높다면, 서비스를 유료로 결제할 의사가 있는 고객들에게 주말에 결제를 유도하는 메일을 보낼 수 있습니다.


## 9. 결제 / 캔슬 / 환불의 총 인원 수와 비율을 구해주세요.
>-  1) 서비스를 유료로 이용중인 사람(completed), 2) 서비스를 더 이상 이용하지 않고 캔슬한 사람(cancelled) / 3) 서비스를 결제했으나 환불한 사람(refunded)의 비율을 알고 싶습니다. 

In [24]:
print(data['Status'].value_counts())
print(data['Status'].value_counts(normalize=True))

completed    5400
cancelled    4010
refunded      590
Name: Status, dtype: int64
completed    0.540
cancelled    0.401
refunded     0.059
Name: Status, dtype: float64



## 10. 성별과 나이별 결제 / 캔슬 / 환불의 총 인원 수와 비율을 구해주세요.
>- age group: 17세 이하
/ 18세 이상, 24세 이하
/ 25세 이상, 35세 이하
/ 36세 이상, 44세 이하
/ 45세 이상, 54세 이하
/ 55세 이상

In [25]:
data['Age(group)'] = pd.cut(data['Age(clean)'], 
                            bins=[0,17,24,35,44,54,99],
                            labels=['00 ~ 17', '18 ~ 24', '25 ~ 35', '36 ~ 44', '45 ~ 54', '55 ~ 99'])

In [26]:
table = pd.pivot_table(data, 
                       index=['Gender(clean)', 'Age(group)'], 
                       values='Age',
                       columns=['Status'],
                       fill_value=0,
                       aggfunc='count',
                       margins=True, margins_name='total')
table['conversion'] = table['completed'] / table['total']
table

Status                    cancelled  completed  refunded  total  conversion
Gender(clean) Age(group)                                                   
female        00 ~ 17            25         35         3     63    0.555556
              18 ~ 24          1637       1827       149   3613    0.505674
              25 ~ 35          1664       2288       271   4223    0.541795
              36 ~ 44           206        421        46    673    0.625557
              45 ~ 54            74        160        25    259    0.617761
              55 ~ 99             0          5         0      5    1.000000
male          00 ~ 17             3          1         0      4    0.250000
              18 ~ 24            80        100        11    191    0.523560
              25 ~ 35           235        404        57    696    0.580460
              36 ~ 44            21         72         9    102    0.705882
              45 ~ 54             9         13         3     25    0.520000
              55 ~ 99             0          1         0      1    1.000000
total                          3954       5327       574   9855    0.540538

>- 분석 결과는 다음과 같습니다.

> * 가장 많은 양의 결제가 일어난 구간은 여성 25 ~ 35세입니다. 총 2288개로, 결제 완료의 40% 이상이 이 구간에서 발생했습니다. 심지어 전환율(conversion)도 54.1%로 평균 이상입니다.
또한 어느 정도 모수가 받쳐주는(결제 완료 100회 이상) 채널 중 이보다 전환율이 높은 채널은 1) 여성 36 ~ 54세, 2) 남성 25 ~ 35세, 3) 남성 36 ~ 44세 입니다. 이 채널들은 전환율이 60% 이상으로 매우 높습니다.
다만 이 채널들의 총 결제자(total)가 낮다는 것은 1) 아직 이 마케팅 채널이 최적화가 덜 되었거나, 2) 고객 획득 비용(CAC)이 높은 편이라 마케팅 비용을 늘리지 않았을 가능성이 있습니다. 또한 아주 희소한 경우이지만, 3) 주 마케팅 채널(ex: 페이스북)에 위 채널에 해당하는 고객의 인원수가 부족할 수도 있습니다.
이런 상황에서, 데이터분석가는 퍼포먼스 마케터와 함께 다음의 아이디어를 제시하여 회사의 매출을 증대할 수 있습니다.

> * 마케팅 예산을 여성 36 ~ 54세쪽에 집중한다. 이 채널이 전환율이 높기 때문에, CAC가 여성 25 ~ 35세와 동일하다면 여성 36 ~ 54세에 마케팅 예산을 늘리는 것은 좋은 전략입니다.
여성 36 ~ 54세 채널의 CAC가 상대적으로 높다면, 이 CAC을 낮추는 시도를 합니다. 이 전략이 성공하면 그 후에 마케팅 예산을 집중하는 것도 방법입니다.
현재 이용하고 있는 광고 채널을 다각화하여, 여성 36 ~ 54세가 활동하는 곳에 집중적으로 마케팅 예산을 투입하는 것도 시도해볼만 합니다.


## 11. 날짜와 요일 / 시간별 결제 / 캔슬 / 환불 비율을 구해주세요.
>- 마케팅팀이 이 정보를 파악할 수 있다면, 1) 전환율이 높은 시기에 마케팅 예산 투입 비중을 줄이고/늘려서 CAC를 낮추거나, 2) 특정 시간대에 눔 코치의 유로 서비스를 아직 구매하지 않은 무료 사용자에게 유료 서비스 구매를 유도하는 메일을 보내서 매출을 늘릴 것입니다.
>- 1) 0시 ~ 23시 사이의 결제/캔슬/환불 비율. 2) 월요일-일요일 사이의 결제/캔슬/환불 비율을 구해주세요.

In [27]:
# 1) 0시 ~ 23시 사이의 결제/캔슬/환불 비율
data['Purchased At(hour)'] = data['Purchased At'].dt.hour

table2 = pd.pivot_table(data,
                        index='Purchased At(hour)',
                        values='Name',
                        columns='Status',
                        aggfunc='count', 
                        margins=True, margins_name='total')
table2['conversion'] = table2['completed'] / table2['total']
table2

Status              cancelled  completed  refunded  total  conversion
Purchased At(hour)                                                   
0                         236        344        26    606    0.567657
1                         156        207        28    391    0.529412
2                          90         97         4    191    0.507853
3                          58         66         5    129    0.511628
4                          59         45         7    111    0.405405
5                          36         47         6     89    0.528090
6                          48         70         6    124    0.564516
7                          80        114        20    214    0.532710
8                         171        264        29    464    0.568966
9                         162        239        36    437    0.546911
10                        208        323        38    569    0.567663
11                        212        263        27    502    0.523904
12                        205        235        36    476    0.493697
13                        205        286        41    532    0.537594
14                        192        253        20    465    0.544086
15                        187        231        14    432    0.534722
16                        187        235        26    448    0.524554
17                        180        246        26    452    0.544248
18                        194        260        20    474    0.548523
19                        163        269        36    468    0.574786
20                        184        236        22    442    0.533937
21                        231        329        32    592    0.555743
22                        248        332        41    621    0.534622
23                        318        409        44    771    0.530480
total                    4010       5400       590  10000    0.540000

>- 분석 결과는 다음과 같습니다.

> - 아쉽게도, 구매 시간별 전환율(conversion)은 큰 차이가 없어 보입니다, 그 의미는 **특정 시간대에 구매한 사용자들이 서비스를 이탈할 확률이 높아지거나 낮아지는 현상은 없다고 볼 수 있습니다.**
다만 전환율과는 별개로, **주로 점심시간(10시 ~ 12시)나 새벽(23시 ~ 24시)에 구매량이 대폭 늘어난다는 것을 알 수 있습니다.** 만일 광고 예산을 집행한다면 이 시기에 집중적으로 집행하거나, 무료 사용자에게 유료 사용자로 전환을 유도하는 이메일을 보냄으로써 전환율을 높이는 것은 시도해볼만 합니다.

In [28]:
# 2) 월요일-일요일 사이의 결제/캔슬/환불 비율
data['Purchased At(weekday)'] = data['Purchased At'].dt.day_name()

table3 = pd.pivot_table(data,
                        index='Purchased At(weekday)',
                        values='Name',
                        columns='Status',
                        aggfunc='count', 
                        margins=True, margins_name='total')
table3['conversion'] = table3['completed'] / table3['total']
table3 = table3.reindex(['Monday','Tuesday','Wednesday','Thursday','Friday','Saturday','Sunday'])
table3

Status                 cancelled  completed  refunded  total  conversion
Purchased At(weekday)                                                   
Monday                       691        863        93   1647    0.523983
Tuesday                      694        935       102   1731    0.540150
Wednesday                    679        953        90   1722    0.553426
Thursday                     616        813        88   1517    0.535926
Friday                       490        674        56   1220    0.552459
Saturday                     412        537        73   1022    0.525440
Sunday                       428        625        88   1141    0.547765

>- 분석 결과는 다음과 같습니다.

> - 구매 시간과 마찬가지로, 구매 요일별 전환율(conversion)은 큰 차이가 없어 보입니다. 어느 요일이나 마찬가지로, 구매한 사람이 서비스를 이탈하거나 남을 확률은 거의 동일합니다.
하지만 사용자들은 전반적으로 **주말(금-일)이 다가올수록 구매를 덜 하게되고, 주중(월-수)이 다가올수록 구매를 많이 하게 되는 현상을 발견할 수 있습니다.** 이 시기에 광고 예산을 크게 집행하거나, 구매를 유도하는 메일이나 모바일 노티피케이션을 보내는 것은 좋은 아이디어입니다.


## 12. 채널별 결제 / 캔슬 / 환불 비율을 알고 싶다.
>- 이 채널별 마케팅 효율 정보를 알 수 있다면, 마케팅 팀에서 마케팅 예산을 재조정하여 1) 마케팅 효율이 좋은 채널에 예산을 집중하고, 2) 반대로 마케팅 효율이 좋지 않은 채널에 예산을 빼는 재조정(rebalancing)을 할 수 있습니다.

In [29]:
data['Channel'].value_counts()

facebook     6880
others       1390
naver        1009
direct        297
email         271
google        120
instagram      33
Name: Channel, dtype: int64

In [30]:
table4 = pd.pivot_table(data,
                        index='Channel',
                        values='Name',
                        columns='Status',
                        aggfunc='count',
                        margins=True, margins_name='total')
table4['conversion'] = table4['completed'] / table4['total']
table4

Status     cancelled  completed  refunded  total  conversion
Channel                                                     
direct           119        169         9    297    0.569024
email             93        155        23    271    0.571956
facebook        2812       3654       414   6880    0.531105
google            42         66        12    120    0.550000
instagram         13         17         3     33    0.515152
naver            386        568        55   1009    0.562934
others           545        771        74   1390    0.554676
total           4010       5400       590  10000    0.540000

>- 이 결과를 통해 알 수 있는 정보는 다음과 같습니다.

> - 현재 가장 많은 구매가 일어나는 채널은 **페이스북(facebook)** 입니다. 거의 대부분의 구매가 이 채널에서 일어났습니다.
구매량이 100회 이상인 채널 중 가장 전환율이 높은 채널은 이메일(email) 입니다. 이 채널은 사용자가 눔의 웹사이트에 방문한 뒤, 바로 구매하지 않고 이메일 주소만만 남겨놨을 경우에 해당됩니다.
아직 구매량이 페이스북만큼 많지는 않지만, 전환률이 페이스북보다 높은 채널 중 하나는 네이버(naver)입니다. 전환율이 56%로 페이스북보다 다소 높은 편입니다.
네이버만큼이나 전환율이 높은 채널은 기타(others)입니다. 이 채널은 결제율이 페이스북만큼 높음에도 불구하고, 아쉽게도 기록이 잘 되어있지 않기 때문에 분석이 어렵습니다.
이 분석 결과를 통해 얻을 수 있는 아이디어는 다음과 같습니다.

> - 먼저 내부에서 트래킹 코드나 데이터 클리닝 코드를 수정하여, **기타(others) 채널을 더 세분화시킬 필요가 있습니다.** 기타 채널은 1) 페이스북 만큼이나 구매량이 많으며, 2) 전환율이 페이스북보다 높습니다. 이 채널을 더 세분화시켜 분석한다면 마케팅 효율을 높일 수 있는 새로운 아이디어가 나올 수 있습니다.
페이스북 다음으로 네이버 검색채널을 집중적으로 튜닝하거나 예산을 배정하여 마케팅 채널을 다각화할 수 있습니다.
이메일(email)로 들어온 사용자가 전환율이 높은 이유를 더 분석할 수 있다면 좋겠습니다. 추측컨데, 눔 코치에 대한 신뢰도를 높일 다양한 정보를 이메일로 수신하였기 때문에 다른 채널에 비해 전환율이 높다는 가설을 세울 수 있습니다. 이 가설이 맞다면, 눔 코치를 이용하는 다른 사용자에게도 동일한 정보를 제공한다면 전체 전환율을 높일 수 있을 것입니다.

----
# (3) 코치 데이터와 매칭 - 코칭팀
>- 코치 데이터 분석에서 가장 중요한 것은, 좋은 코칭을 하는 사람과 그렇지 못한 사람을 구분하는 것입니다.
코칭팀에서는 좋은 코칭을 하는 코치의 노하우를 정리하여 다른 코치들에게 전파할 필요가 있고, 정 반대로 좋지 않은 코칭을 하는 코치와는 개별 면담을 통해 코칭 퀄리티를 높여야 합니다. 좋은 코칭과 좋지 않은 코칭은 결제 비율과 캔슬 비율, 그리고 환불 비율로 판단할 수 있습니다.

## 13. 기존의 데이터와 코치 데이터를 합쳐주세요.

In [51]:
#coach = coach.set_index('Access Code')
data3 = pd.merge(data[['Name','Status']], coach, on=['Access Code'], how='left')
data3.head()

Access Code Name     Status  정은오 코치(VEV4PGJB)  오승혁 코치(VENPKBP9)  \
0    Y9RY2VSI  김승혜  completed                 0                 0   
1    3GTN3S3B  허승준  completed                 0                 0   
2    6B0IG276  이지민  completed                 0                 0   
3    EMGRU2MO  장설윤  completed                 0                 0   
4    1ELG96TX  서성빈  completed                 0                 0   

   조소은 코치(D0WASBXN)  고영재 코치(C91AVNGB)  조수민 코치(OBCAO3W0)  강채아 코치(WH2NIKCO)  \
0                 0                 0                 0                 0   
1                 0                 0                 0                 0   
2                 0                 0                 0                 0   
3                 0                 0                 0                 0   
4                 0                 0                 0                 0   

   황다훈 코치(1I6IWURH)  ...  오초빈 코치(A3WOLAQM)  서수정 코치(F36LORFC)  \
0                 0  ...                 0                 0   
1                 0  ...                 0                 0   
2                 0  ...                 0                 0   
3                 1  ...                 0                 0   
4                 0  ...                 0                 0   

   정서율 코치(LX1G7EMD)  고우재 코치(SKNL9Z4P)  문한규 코치(OU1WVDGA)  황세안 코치(3QUBQAVE)  \
0                 0                 1                 0                 0   
1                 0                 0                 1                 0   
2                 0                 0                 0                 0   
3                 0                 0                 0                 0   
4                 0                 0                 0                 0   

   홍성은 코치(2I3QJQ5O)  고성은 코치(34T7XPYR)  백한율 코치(HPWAN8R0)  안슬은 코치(QAVWJSZ1)  
0                 0                 0                 0                 0  
1                 0                 0                 0                 0  
2                 0                 0                 0                 0  
3                 0                 0                 0                 0  
4                 0                 0                 0                 0  

[5 rows x 103 columns]


## 14. 코치별 담당 사용자(total) / 구매 완료 횟수(completed) / 캔슬 횟수(canceled) / 환불 횟수(refunded)를 구해주세요.

In [52]:
data3 = data3.reset_index().melt(id_vars=['Access Code','Name','Status'])
data3 = data3[data3['value'] != 0]

table5 = pd.pivot_table(data3,
                        index='variable',
                        values='value',
                        columns='Status',
                        aggfunc='sum',
                        fill_value=0,
                        margins=True, margins_name='total')
table5

Status            cancelled  completed  refunded     total
variable                                                  
index              20227915   26772885   2994200  49995000
강은우 코치(EJIHL7OE)        122        171        16       309
강지희 코치(NOEP7X8B)         15         19         1        35
강채아 코치(WH2NIKCO)          9         10         2        21
고성은 코치(34T7XPYR)         24         36         1        61
...                     ...        ...       ...       ...
황세안 코치(3QUBQAVE)         48         61         8       117
황소영 코치(91YZ8NY0)          5         13         2        20
황시준 코치(M7EJJXFT)         16         16         2        34
황재우 코치(YIFMV1GQ)          7          6         1        14
total              20233198   26779850   2995009  50008057

[101 rows x 4 columns]


## 15. 코치별 전환율(conversion rate) / 취소율(cancellation rate)를 계산해주세요.
>- 전환율 = 구매 완료(completed)를 한 사람 / 전체 구매자
>- 취소율 = 취소(cancelled)나 환불(refunded)을 한 사람 / 전체 구매자

>- 1) 전환율이 높은 코치, 2) 취소율이 높은 코치 순으로 정렬해주세요. 단 모수가 부족한 경우를 배제하기 위해, 코칭을 100회 이상 하지 않은 사용자는 배제하도록 하겠습니다.

In [53]:
# 전환율
table5['conversion_rate'] = table5['completed'] / table5['total']
# 취소율
table5['cancellation_rate'] = (table5['cancelled'] + table5['refunded']) / table5['total']
table5 = table5[table5['total'] >= 100]

In [54]:
# 1) 전환율이 높은 코치
table5.sort_values(['conversion_rate'], ascending=False).head()

Status            cancelled  completed  refunded  total  conversion_rate  \
variable                                                                   
조우찬 코치(WWN531JQ)         36         65         6    107         0.607477   
허슬지 코치(DWVG5IFL)         43         71         3    117         0.606838   
허성원 코치(9124O1IH)         43         76         7    126         0.603175   
조설영 코치(U7L98DAO)         48         78         6    132         0.590909   
권슬영 코치(E3GD4106)         42         70         9    121         0.578512   

Status            cancellation_rate  
variable                             
조우찬 코치(WWN531JQ)           0.392523  
허슬지 코치(DWVG5IFL)           0.393162  
허성원 코치(9124O1IH)           0.396825  
조설영 코치(U7L98DAO)           0.409091  
권슬영 코치(E3GD4106)           0.421488

In [55]:
# 2) 취소율이 높은 코치
table5.sort_values(['cancellation_rate'], ascending=False).head()

Status            cancelled  completed  refunded  total  conversion_rate  \
variable                                                                   
조수민 코치(OBCAO3W0)         49         46         8    103         0.446602   
박도영 코치(I4KVQ5G0)         77         70         6    153         0.457516   
오동완 코치(0O48DQCH)         55         56         7    118         0.474576   
조초연 코치(3JBE9GKO)         55         56         5    116         0.482759   
오초빈 코치(A3WOLAQM)        124        150        27    301         0.498339   

Status            cancellation_rate  
variable                             
조수민 코치(OBCAO3W0)           0.553398  
박도영 코치(I4KVQ5G0)           0.542484  
오동완 코치(0O48DQCH)           0.525424  
조초연 코치(3JBE9GKO)           0.517241  
오초빈 코치(A3WOLAQM)           0.501661